core/src/test/java/org/apache/calcite/test/enumerable/EnumerableRepeatUnionTest.java

```diff
@@ -45,6 +45,17 @@
  */
 class EnumerableRepeatUnionTest {
 
+  @Test void testGenerateNumbersUsingSql() {
+    CalciteAssert.that()
+        .query("WITH RECURSIVE delta(n) AS (\n"
+            + "  VALUES (1)\n"
+            + " UNION ALL\n"
```

Comment:
shouldn't UNION ALL actually never converge?
Why isn't the result collection a multiset that grows infinitely?

Reply:
@mihaibudiu the `WHERE n < 10` breaks the "infinite growth". We get the same behavior [in e.g. PostgreSQL](http://sqlfiddle.com/#!17/9eecb/111270)

Reply:
The `WHERE` breaks the growth in the value of n, but UNION would break the growth in cardinality. 
UNION ALL should produce a multiset, so you should get the following values in `delta` in successive iterations: [1], [1, 2], [1, 2, 2, 3], [1, 2, 2, 2, 3, 3, 4] etc in your relation in successive iterations. I understand how this should work with `UNION`, but not with `UNION ALL`. Why is delta a set and not a multiset?

Reply:
@mihaibudiu  Thanks for reviewing the changes. I agree with @rubenada regarding the explanation as to why this use case doesn't run into infinite growth. I believe in general that semantics of UNION / UNION ALL are subtle in Recursive queries. 

Here is an example (again from postgresql) where usage of the union all (http://sqlfiddle.com/#!17/9eecb/111278) will run into infinite loop but usage of union works fine.
The infinite loop for the above example actually is because of the result set being a multi set when union all is used.

Reply:
More info about how RECURSIVE UNION [ALL] works internally can be found here:
https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-RECURSIVE
Calcite implementation was heavily inspired by this postgresql description.